runtime/src/accounts_db.rs

```diff
@@ -3249,6 +3249,172 @@ impl AccountsDb {
         total_accounts_after_shrink
     }
 
+    /// Remove all accounts not in `minimized_account_set`. This should only be used by ledger-tool to create minimized snapshots.
+    pub(crate) fn minimize_accounts_db(
+        &self,
+        snapshot_slot: Slot,
+        minimized_account_set: &DashSet<Pubkey>,
+    ) {
+        let minimized_slot_set = self.get_minimized_slot_set(minimized_account_set);
+
+        let mut dead_slots = Vec::new();
+        let mut dead_storages = Vec::new();
+        let mut count = 0;
+        let mut filter_storages_us = 0;
+
+        let snapshot_storages = self.get_snapshot_storages(snapshot_slot, None, None).0;
+        let num_slots = snapshot_storages.len();
+
+        for storages in snapshot_storages {
+            let slot = storages.first().unwrap().slot();
+            if slot == snapshot_slot {
+                continue;
+            }
+            if minimized_slot_set.contains(&slot) {
+                let (_, filter_storages_measure) = measure!(
+                    self.filter_storages(storages, minimized_account_set, &mut dead_storages),
+                    "filter storages"
+                );
+                filter_storages_us += filter_storages_measure.as_us();
+            } else {
+                dead_slots.push(slot);
+            }
+
+            count += 1;
+            if count % 1000 == 0 {
+                info!(
+                    "filtering storages took {filter_storages_us}us progress {count}/{num_slots}"
+                );
+                filter_storages_us = 0;
+            }
+        }
+
+        let (_, purge_slots_measure) = measure!(
+            self.purge_slots_from_cache_and_store(
+                dead_slots.iter(),
+                &self.external_purge_slots_stats,
+                true,
+            ),
+            "purge slots"
+        );
+        info!("{purge_slots_measure}");
+
+        let (_, drop_or_recycle_stores_measure) = measure!(
+            self.drop_or_recycle_stores(dead_storages),
+            "drop or recycle stores"
+        );
+        info!("{drop_or_recycle_stores_measure}");
+    }
+
+    fn get_minimized_slot_set(&self, minimized_account_set: &DashSet<Pubkey>) -> DashSet<Slot> {
+        let (minimized_slot_set, minimized_slot_set_measure) = measure!(
+            {
+                let minimized_slot_set = DashSet::new();
+                minimized_account_set.par_iter().for_each(|pubkey| {
+                    if let Some(read_entry) = self.accounts_index.get_account_read_entry(&pubkey) {
+                        read_entry.slot_list().iter().for_each(|(slot, _)| {
```

Comment:
This will add the account at every slot it exists in. We only want the highest slot #. We can assume this always runs when all slots are roots since we're starting from a snapshot. This can be done with ```
if let Some(max_slot) = iter().map().max() { .insert(max_slot)}

Reply:
the way you did it isn't incorrect, just not minimal